### ISSUE TITLE
Allow padding on simple returns for newline-before-return

### ISSUE THREAD

**Issue by andrewvaughan** [eslint/eslint#6176]
**ESLint Version:** 2.10.1

## Problem:
Using `newline-before-return` with `padded-blocks` causes a conflict, as `padded-blocks` insists on padded spacing for functions, and `newline-before-return` complains if there is a newline before a return statement by itself in a function.

## Recommended Solution:
Add an option to `newline-before-return` to ignore scenarios where the only statement in a function block is the return statement.

For example:

```js
"newline-before-return": ["error", {
    "pad-getters": true  // (New) Ensure getter returns are padded on either side by a newline
}]
```

This will allow ESLint to properly support it's built-in plugins without changing the default functionality of the `newline-before-return` rule.  With the above configuration:

**This code would result in an error:**
```js
function getVar() {
    return this.var;
}
```

**This code would be valid with both plugins:**
```js
function getVar() {

    return this.var;

}
```

**Comment by alberto** [comment]
I think `newline-before-return` should not error when it is the first statement in a block, regardless of the newlines and that should be left to `padded-blocks`.

The doc says (emphasis mine):

> This rule requires an empty line before `return` statements to increase code clarity, except when the `return` is alone inside a statement group (such as an if statement). In the latter case, the `return` statement **does not need** to be delineated by virtue of it being alone.

**Comment by kaicataldo** [comment]
I support @alberto's suggested change :+1:

**Comment by mysticatea** [comment]
I added `accepted` label since there are 4 :+1:s from core team.

**Comment by alberto** [comment]
There is still the question of whether this is a breaking change and it should instead be treated as an enhancement go behind an option @eslint/eslint-team

**Comment by mysticatea** [comment]
I think we can modify the default behavior because this change is relaxing, so it probably will not break existing CI tests of users.

**Comment by mikesherov** [comment]
In JSCS, relaxing rules was not considered a breaking change.

**Comment by markelog** [comment]
Can we document what should be considered as breaking change? See JSCS example https://github.com/jscs-dev/node-jscs/blob/master/OVERVIEW.md#user-content-versioning--semver for reference

**Comment by nzakas** [comment]
I concur that this is a bug as the documentation seems to indicate that the desired behavior.

I've opened an issue so we can discuss formalizing a semver policy: https://github.com/eslint/eslint/issues/6244

**Comment by kalisjoshua** [comment]
I like this as an option. When should this be available? It looks like it is already included but I am not seeing it working.

**Comment by kaicataldo** [comment]
Hey there, this should be in as of >2.11.0 http://eslint.org/blog/2016/05/eslint-v2.11.0-released#bug-fixes

**Comment by kalisjoshua** [comment]
@kaicataldo That's what I was thinking but I am not able to get this option to work; I still get:

```
error  Block must not be padded by blank lines  padded-blocks
```

... in my lint results output.

My `.eslintrc.json` looks like this:

```
{
  "env": {
    "browser": true,
    "node": true
  },
  "extends": "airbnb",
  "rules": {
    "comma-dangle": ["warn", "never"],
    "consistent-return": "error",
    "eqeqeq": ["error", "smart"],
    "newline-before-return": "off",
    "no-use-before-define": ["error", {
      "functions": false
    }],
    "space-before-function-paren": ["warn", "always"]
  }
}
```

I've tried all permutation of `newline-before-return` and `padded-blocks`. The only thing that seems to be working is `padded-blocks` which affects everything and I only want to allow an additional line before `return` statements. Am I understanding this correctly?

**Comment by kaicataldo** [comment]
Ah yeah, that's the `padded-blocks` rule warning, not the `newline-before-return` rule. I recommend making a new issue for this, since this issue is unrelated and closed. Please make sure you include the info asked for in the template, including the code that's throwing the error. Thanks!

**Comment by kalisjoshua** [comment]
@kaicataldo Thank you, I will.

**Comment by kaicataldo** [comment]
Looks like this fix was just for newlines before `return` statements. The `padded-blocks` warning is for the missing newline after the `return` statement (on line 4 in your example). The change in this issue makes the rules compatible by relaxing the `newline-before-return` rule to allow a newline before the `return` statement even when it's the only statement in the block. It sounds like you might not want to be using the `padded-blocks` rule, if you don't want a padding newline before the end of the block.
